directorywatcher/directorywatcher.go

```diff
@@ -0,0 +1,288 @@
+) {
+	for {
+		select {
+		case <-r.ctx.Done():
+			watcher.Close()
+			return
+
+		case err := <-watcher.Errors:
+			logger.Log(context.Background(), "directorywatcher: watcher error: "+err.Error())
+
+		case ev := <-watcher.Events:
+
+			isDir, _ := isDirectory(ev.Name)
+
+			switch ev.Op {
+			default:
+				// do nothing.
+			case fsnotify.Create: // add to watcher, parse if file
+				// Wrap with an anonymous function to make sure that defer works.
+				func() {
+					if isDir {
+						watcher.Add(ev.Name)
+					} else {
+						f, err := limitopen.OpenWithLimit(ev.Name, softLimit, hardLimit)
+						if err != nil {
+							logger.Log(context.Background(), "directorywatcher: I/O error: "+err.Error())
+							return
+						}
+						defer f.Close()
+						d, err := parser(f)
+						if err != nil {
+							logger.Log(context.Background(), "directorywatcher: parser error: "+err.Error())
+						} else {
+							data := r.data.Load()
+							data, err = add(data, d)
+							if err != nil {
+								logger.Log(context.Background(), "directorywatcher: add file error: "+err.Error())
+								return
+							}
+							r.data.Store(data)
+						}
+					}
+				}()
+			case fsnotify.Rename, fsnotify.Remove: // remove
+				// Wrap with an anonymous function to make sure that defer works.
+				func() {
+					// remove data related to file
+					data := r.data.Load()
+					data, err := remove(data, ev.Name)
+					if err != nil {
+						logger.Log(context.Background(), "directorywatcher: remove file error: "+err.Error())
+						return
+					}
+					r.data.Store(data)
+				}()
+			case fsnotify.Write, fsnotify.Chmod: //parse
+				// Wrap with an anonymous function to make sure that defer works.
+				func() {
+					if isDir {
+						// do nothing
+					} else {
+						f, err := limitopen.OpenWithLimit(ev.Name, softLimit, hardLimit)
+						if err != nil {
+							logger.Log(context.Background(), "directorywatcher: I/O error: "+err.Error())
+							return
+						}
+						defer f.Close()
+						d, err := parser(f)
+						if err != nil {
+							logger.Log(context.Background(), "directorywatcher: parser error: "+err.Error())
+						} else {
+							data := r.data.Load()
+							data, err = add(data, d)
+							if err != nil {
+								logger.Log(context.Background(), "directorywatcher: add file error: "+err.Error())
+								return
+							}
+							r.data.Store(data)
+						}
+					}
+				}()
+			}
+		}
+	}
+}
+
+var (
+	_ DirectoryWatcher = (*Result)(nil)
+)
+
+// New initializes a directorywatcher designed for recursivly
+// looking through a directory instead of a file
+func New(ctx context.Context, cfg Config) (*Result, error) {
+	limit := cfg.MaxFileSize
+	if limit <= 0 {
+		limit = filewatcher.DefaultMaxFileSize
+	}
+	hardLimit := limit * filewatcher.HardLimitMultiplier
+
+	var watcher *fsnotify.Watcher
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	var d interface{}
+	res := &Result{}
+	// res.data.Store(Folder{
```

Comment:
don't comment out code; either delete it, or `if false {}` it or something if you will need it in the future.